### ISSUE TITLE
Sarama completely hangs(doesn't send) somewhere when a broker goes down in case of using DNS url

### ISSUE THREAD

**Issue by safiqueahemad** [Shopify/sarama#661]
Hi, 

I am using mesos url (similar to url resolved by DNS server) to deploy Kafka cluster.
When a broker goes down then client doesn't send any message for Net DialTimeout period.
If this DialTimeout is reduced then problem also reduced accordingly.

Here problem is, client should continue sending message to other broker which are completely unaffected. 

Kindly, help me why it is completely not sending message for DialTimeout which is by default set to 30  Second.

**Comment by eapache** [comment]
What version of Sarama are you running? If this is not the latest version, please try updating.

You can set `sarama.Logger` to a `log.Logger` instance to capture debug output. Please provide these logs from when the issue occurs.

Are you setting any non-default configuration values? If so, which ones and to what values?

---

Off the top of my head, this could just be the channels backing up: when a broker hangs but you try and produce messages to it, we have to store those messages locally in memory until the broker recovers. If that buffer fills up, Sarama will stop accepting new messages entirely since it can't tell immediately which broker a message should go to. If this is the case you can try increasing `ChannelBufferSize`.

**Comment by safiqueahemad** [comment]
[Sarama] 2016/05/18 07:54:47 producer/broker/3 state change to [closing] because write tcp 10.19.15.181:53512->10.3.13.129:9092: write: connection reset by peer
[Sarama] 2016/05/18 07:54:47 Closed connection to broker kafka3-dev9-dakota.mon-marathon-service.mesos:9092
[Sarama] 2016/05/18 07:54:47 producer/leader/safreliability/12 state change to [retrying-1]
[Sarama] 2016/05/18 07:54:47 producer/leader/safreliability/12 abandoning broker 3
[Sarama] 2016/05/18 07:54:47 client/metadata fetching metadata for [safreliability] from broker kafka3-dev9-dakota.mon-marathon-service.mesos:9092
[Sarama] 2016/05/18 07:54:47 client/metadata got error from broker while fetching metadata: EOF
[Sarama] 2016/05/18 07:54:47 Closed connection to broker kafka3-dev9-dakota.mon-marathon-service.mesos:9092
[Sarama] 2016/05/18 07:54:47 client/metadata fetching metadata for [safreliability] from broker kafka2-dev9-dakota.mon-marathon-service.mesos:9092
[Sarama] 2016/05/18 07:54:47 Connected to broker at kafka2-dev9-dakota.mon-marathon-service.mesos:9092 (unregistered)
[Sarama] 2016/05/18 07:54:47 producer/broker/3 starting up
[Sarama] 2016/05/18 07:54:47 producer/leader/safreliability/12 selected broker 3
[Sarama] 2016/05/18 07:54:47 producer/broker/3 state change to [open] on safreliability/12
[Sarama] 2016/05/18 07:54:47 producer/leader/safreliability/0 state change to [retrying-1]
[Sarama] 2016/05/18 07:54:47 producer/leader/safreliability/0 abandoning broker 3
[Sarama] 2016/05/18 07:54:47 client/metadata fetching metadata for [safreliability] from broker kafka2-dev9-dakota.mon-marathon-service.mesos:9092
[Sarama] 2016/05/18 07:54:47 producer/leader/safreliability/0 selected broker 3
[Sarama] 2016/05/18 07:54:47 producer/broker/3 state change to [open] on safreliability/0
[Sarama] 2016/05/18 07:54:47 producer/leader/safreliability/3 state change to [retrying-1]
[Sarama] 2016/05/18 07:54:47 producer/leader/safreliability/3 abandoning broker 3
[Sarama] 2016/05/18 07:54:47 client/metadata fetching metadata for [safreliability] from broker kafka2-dev9-dakota.mon-marathon-service.mesos:9092
[Sarama] 2016/05/18 07:54:47 producer/leader/safreliability/3 selected broker 3
[Sarama] 2016/05/18 07:54:47 producer/broker/3 state change to [open] on safreliability/3
[Sarama] 2016/05/18 07:54:47 producer/leader/safreliability/9 state change to [retrying-1]
[Sarama] 2016/05/18 07:54:47 producer/leader/safreliability/9 abandoning broker 3
[Sarama] 2016/05/18 07:54:47 client/metadata fetching metadata for [safreliability] from broker kafka2-dev9-dakota.mon-marathon-service.mesos:9092
[Sarama] 2016/05/18 07:54:47 producer/leader/safreliability/9 selected broker 3
[Sarama] 2016/05/18 07:54:47 producer/broker/3 state change to [open] on safreliability/9
[Sarama] 2016/05/18 07:54:47 producer/leader/safreliability/6 state change to [retrying-1]
[Sarama] 2016/05/18 07:54:47 producer/leader/safreliability/6 abandoning broker 3
[Sarama] 2016/05/18 07:54:47 producer/broker/3 shut down
[Sarama] 2016/05/18 07:54:47 client/metadata fetching metadata for [safreliability] from broker kafka2-dev9-dakota.mon-marathon-service.mesos:9092
[Sarama] 2016/05/18 07:54:47 producer/leader/safreliability/6 selected broker 3
[Sarama] 2016/05/18 07:54:47 producer/broker/3 state change to [open] on safreliability/6
[Sarama] 2016/05/18 07:54:47 producer/leader/safreliability/12 state change to [flushing-1]
[Sarama] 2016/05/18 07:54:47 producer/leader/safreliability/0 state change to [flushing-1]
[Sarama] 2016/05/18 07:54:47 producer/leader/safreliability/3 state change to [flushing-1]
[Sarama] 2016/05/18 07:54:47 producer/leader/safreliability/9 state change to [flushing-1]
[Sarama] 2016/05/18 07:54:47 producer/leader/safreliability/6 state change to [flushing-1]
[Sarama] 2016/05/18 07:54:47 producer/leader/safreliability/12 state change to [normal]
[Sarama] 2016/05/18 07:54:47 producer/leader/safreliability/3 state change to [normal]
[Sarama] 2016/05/18 07:54:47 producer/leader/safreliability/0 state change to [normal]
[Sarama] 2016/05/18 07:54:47 producer/leader/safreliability/9 state change to [normal]
[Sarama] 2016/05/18 07:54:47 producer/leader/safreliability/6 state change to [normal]
**[Sarama] 2016/05/18 07:54:47 producer/broker/3 maximum request accumulated, waiting for space
[Sarama] 2016/05/18 07:55:17 Failed to connect to broker kafka3-dev9-dakota.mon-marathon-**service.mesos:9092: dial tcp 10.3.13.129:9092: i/o timeout
[Sarama] 2016/05/18 07:55:17 producer/broker/3 state change to [closing] because dial tcp 10.3.13.129:9092: i/o timeout
[Sarama] 2016/05/18 07:55:17 producer/broker/3 state change to [closing] because dial tcp 10.3.13.129:9092: i/o timeout
[Sarama] 2016/05/18 07:55:17 producer/leader/safreliability/12 state change to [retrying-2]
[Sarama] 2016/05/18 07:55:17 producer/leader/safreliability/12 abandoning broker 3
[Sarama] 2016/05/18 07:55:17 client/metadata fetching metadata for [safreliability] from broker kafka2-dev9-dakota.mon-marathon-service.mesos:9092
[Sarama] 2016/05/18 07:55:17 producer/broker/3 starting up

-----------------------------------------------------------------------------------------------------------------
There are three brokers 1,2 and 3. Sarama client is producing them.
The above log show since broker 3 was taken down.
The highlighted lines shows the gap of 30 second when nothing happens.
On code and log analysis, it seems that code blocks at channel bp.responses in waitforspace function.
The block happens for 30 sec for down broker 3 because of dial timeout config.
Why it's not able to produce on broker 1 and 2 which are active.

**Comment by eapache** [comment]
These logs verify my original hypothesis. We are running out of space to buffer the messages destined for broker 3, and so cannot accept more messages on *any* broker. You may be able to avoid the problem by simply changing the config (`ChannelBufferSize`, `Producer.Flush`,  potentially others if you're using compression) to provide more buffer space.

Alternative behaviours would be:
- immediately return errors for messages on broker 3; this would break the ordering/retry guarantee but permit other brokers to continue working
- provide direct-to-broker channels for manually partitioned messages (this would not work for any partitioners other than `ManualPartitioner`)

**Comment by safiqueahemad** [comment]
Playing with ChannelBufferSize didn't help.
Below is some more observation. Kindly correct me.
- My understanding is, there is separate go routine to send/receive message with each broker in parallel. Below is function that handles it: 
    func (bp *brokerProducer) run()
- When overflow condition occur then broker enters in waitforspace condition. If there is no broker failure i.e in normal situation then client keep sending message to other brokers. But when a broker goes down then this problem happens. 
Each client maintains broker producer channel with each broker thats is bp.responses. client get hung here in bp.responses until unless timeout happens.
- As said above why broker 3 go routine blocks itself and rest of go routine as well.

**Comment by eapache** [comment]
> why broker 3 go routine blocks itself and rest of go routine as well

Because we don't know in `func (p *asyncProducer) dispatcher()` which broker the message is intended for. The events happen as follows:

- broker 3 goes down
- the goroutine for broker 3 buffers messages while it waits for a response (or a timeout)
- the goroutine for broker 3 runs out of buffer space and blocks (`waitForSpace`)
- some partitionProducer (e.g. for safreliability/12) tries to send another message to broker 3, and blocks
- some topicProducer (e.g. for safreliability) tries to send another message to the partitionProducer, and blocks
- `func (p *asyncProducer) dispatcher()` tries to send another message to the topicProducer, and blocks
- nothing is reading from the input channel because the dispatcher is blocked

**Comment by safiqueahemad** [comment]
Hey Evan,
Thanks for quick analysis and feedback.
I see one more thing here.
When a broker goes down then first metadata retry happens quickly...
Second retry happens after 1 sec and
Third retry happen after quite a long after near 20 second..
here metadata is still not updated from Kafka..
after this messages start failing...
after quite long near 32 second metadata refresh is tried and updated meta data is found.

Due to this so long gaps, if kafka correct its metadata but we fetch it quite late.

**Comment by JarenGlover** [comment]
I was wondering in general is there any risks of using a mesos url (consul in my case)? 

Does the client try to get the entire cluster's metadata from the first IP from the DNS request?
- if the first IP fails doesn't it try the next IP in the DNS list? 
- does it use UDP or TCP to get list of IPs? This can be the diff in getting 3 or all the IPs of a cluster.

**Comment by eapache** [comment]
@JarenGlover: Kafka is 100% TCP. Sarama will walk through the list of setup addresses it is given until it finds one that works. I know nothing about Golang's DNS stack or how Mesos resolves URLs, but Sarama doesn't do anything to try multiple DNS resolutions, so it's quite plausible that if multiple IPs are returned we ignore all but the first. For future questions like this please file a separate ticket, while they are related to this report they are not the same precise issue (and it is much easier to merge accidental duplicates than to split multiple issues apart).

**Comment by eapache** [comment]
@safiqueahemad the metadata fetches are not failing or retrying; they are succeeding but Kafka is returning an indication that the dead broker is still the leader. Note the following:

> [Sarama] 2016/05/18 07:54:47 client/metadata fetching metadata for [safreliability] from broker kafka2-dev9-dakota.mon-marathon-service.mesos:9092
> [Sarama] 2016/05/18 07:54:47 producer/leader/safreliability/3 selected broker 3

This means that the metadata request succeeded and returned that the leader is broker 3 (which failed). Thus Sarama retries to send to broker 3, which blocks until the network timeout occurs. Making kafka correct its metadata sooner depends on configuration of kafka and zookeeper, not on Sarama.

**Comment by safiqueahemad** [comment]
Hey Evan,

For the hang period client get blocked, don't you think its violation of async nature. User calling Send method will also get blocked as channel capacity become full. It will never return until timeout and down brokerProducer again retry message to feed it back.

**Comment by wvanbergen** [comment]
We have to limit the amount of data in channels. As a practical matter, because go only supports channels with a limited capacity. But also by design: if we don't limit it, memory usage can grow without limits, which would mean that eventually you will run into unpredictable results.

Sarama is quite flexible in how it behaves around this scenario: you can shorten timeouts, change how often it flushes, or increase buffer sizes. Potentially, we could add a behavior that it immediately fails messages when the channels are full, instead of blocking. Not sure if that would be a good idea.

**Comment by eapache** [comment]
When this case occurs we have three choices I can see:
- crash from running out of memory
- block (violating async promise)
- return errors (violating ordering/delivery promise)

If there is a fourth option, please say what it is.

**Comment by billygout** [comment]
@eapache
When the AsyncProducer "input" channel is filled up, due to downed brokers on the other side or brokers not keeping up, a select statement with two cases on the two channels "input" and "errors"  would block indefinitely. This is problematic, when you have a "main loop" like this:
```
Loop:
  Receive a message from a UDP port
  Select statement with cases for "errors" and "input" channels into which the received message is sent.
```
The problem with the above is that, since the select statement blocks indefinitely when "input" channel is full, the program will not get around to receiving from the UDP socket in time before UDP messages are dropped. This UDP case is just an example, but the same logic would apply for any application which gathers data from a time-sensitive source and wants to forward to Kafka with the AsyncProducer.

In my case, what I want to do is take any message that fails to be sent and "spill" it to disk to be later attempted to transmit it again.

One solution I have thought of is to add a third case to the select statement:
`case <-time.After(time.Second * 3):`

With this added case, I can guarantee that the loop runs with a maximum delay, so that I won't go without calling "receive" on the UDP socket for longer than I can without dropping incoming packets.

For handling the "spilling" to disk feature, I would write the message to disk in either of the two cases for (1) `time.After` and (2) `producer.Errors()`.

However, I think it would be nice if the AsyncProducer could be configured in such a way that the "Input()" channel would never block, and if the queue were full, the AsyncProducer would immediately return the same message back on the "Errors()" channel. If that were the case, I wouldn't need a `time.After` case in my select statement, and I could handle the "spilling" to disk feature only in the "Errors()" case.

I would be curious to hear whether this would be doable, and if you have any other ideas on how best to handle this use case.

Sorry for the long post.
Thanks!

**Comment by billygout** [comment]
I just realised the loop won't really work anyway, since if I receive and error in Errors() channel, I won't be sending in the Input() channel in the same iteration, but I think the general indefinite-time blocking can still be an issue.

**Comment by eapache** [comment]
> it would be nice if the AsyncProducer could be configured in such a way that the "Input()" channel would never block, and if the queue were full, the AsyncProducer would immediately return the same message back on the "Errors()" channel.

This is not as trivial as it sounds - due to the nature of Go's scheduling, it's entirely possible and reasonable for that particular buffer to fill up and block momentarily while a different goroutine is running internally. We really don't want to start returning errors in that case, which means we'd need a timer with some configurable value (a few seconds probably). But this is exactly what you have already suggested yourself with the `time.After` case, so it doesn't save you any work building it into Sarama.

**Comment by billygout** [comment]
@eapache you are right, and thanks for the response!

I have come up with another solution that seems more idiomatic go style:
